### ISSUE TITLE
`torch.compile`  produce wrong result in backward ad with  `conv2d + interpolate` when  interpolate`mode=nearest/bilinear/bicubic`

### ISSUE THREAD

**Issue by FatPigeorz** [pytorch/pytorch#100794]
### 🐛 Describe the bug

torch.compile leads to wrong results in the following program. Please see the comments for the results of some variants.
~~~ python
import torch
import torch.nn as nn
import copy

torch.manual_seed(0)
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=4)

    def forward(self, x):
        v2 = self.conv(x)
        ret = torch.nn.functional.interpolate(v2, size=[1, 1], mode="nearest")
        # ✅ with mode = 'area', 'nearest-exact'
        # ❌ with mode = 'nearest', 'bilinear', 'bicubic', produce the same wrong result in this program
        return ret

x = torch.ones(size=[1, 1, 4, 4], dtype=torch.float32)
model = Model()
model_tmp = copy.deepcopy(model)

model_tmp.train()
compiled = torch.compile(model_tmp, fullgraph=True)
params = {k: v for k, v in compiled.named_parameters()}
ret_compiled = compiled(x)
grad_compiled = torch.autograd.grad(
                    outputs=ret_compiled,
                    inputs=params.values(),
                    grad_outputs=torch.ones_like(ret_compiled),
                    retain_graph=True,
                    allow_unused=True,
                )

model.train()
params = {k: v for k, v in model.named_parameters()}
ret_eager = model(x)
grad_eager = torch.autograd.grad(
                    outputs=ret_eager,
                    inputs=params.values(),
                    grad_outputs=torch.ones_like(ret_eager), 
                    retain_graph=True,
                    allow_unused=True,
                )

torch.testing.assert_close(ret_compiled, ret_eager)
# ✅ direct output is correct
torch.testing.assert_close(grad_compiled, grad_eager, rtol=1e-2, atol=1e-3)
print("========= backward AD OK! ===========")
# ❌ the gradient calculated with backward ad is wrong

# Mismatched elements: 16 / 16 (100.0%)
# Greatest absolute difference: 1.0 at index (0, 0, 0, 0) (up to 0.001 allowed)
# Greatest relative difference: 1.0 at index (0, 0, 0, 0) (up to 0.01 allowed)
# grad_compiled =  
# (tensor([[[[0., 0., 0., 0.],
#           [0., 0., 0., 0.],
#           [0., 0., 0., 0.],
#           [0., 0., 0., 0.]]]]), tensor([0.]))
#
# grad_eager = 
# (tensor([[[[1., 1., 1., 1.],
#           [1., 1., 1., 1.],
#           [1., 1., 1., 1.],
#           [1., 1., 1., 1.]]]]), tensor([1.]))
#
~~~

### Versions

<details><summary><i>Click to expand.</i></summary>
<div>

PyTorch version: 2.1.0.dev20230505+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2080 Ti
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230505+cpu
[pip3] triton==2.0.0
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.1.0.dev20230505+cpu          pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi
</div>
</details>

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @anijain2305

**Comment by soulitzer** [comment]
Maybe reproducible on CUDA

**Comment by kta-intel** [comment]
this reproduces on cuda:
```
Tesla T4

grad_compiled =  
(tensor([[[[0., 0., 0., 0.],
           [0., 0., 0., 0.],
           [0., 0., 0., 0.],
           [0., 0., 0., 0.]]]], device='cuda:0'),
 tensor([0.], device='cuda:0')

grad_eager = 
tensor([[[[1., 1., 1., 1.],
           [1., 1., 1., 1.],
           [1., 1., 1., 1.],
           [1., 1., 1., 1.]]]], device='cuda:0'),
 tensor([1.], device='cuda:0')
```

**Comment by ezyang** [comment]
As of May 23 fcf812c35ac2eb59bf409e5276fa685f544d435e the repro script seems to pass. Please reopen if upgrading does not help.
